### PR TITLE
establish working definition of group

### PR DESCRIPTION
**Pull request by MAO3J1m0Op**
These will be the working definitions and notations that we will use for the entirety of the project. Chapter 1 should build up to these definitions, possibly using definitions of their own, while chapters 2 and onward should use these definitions in future theorems.

The namespace in which these definitions are located is meant to be temporary, as it will be moved into a chapter once the structure of the project becomes more concrete.

Once we agree, let's merge this PR so that it's only one commit, sqaushing it if necessary.

### CODE REVIEW

File: Main.lean
Comment (MithicSpirit) [inline Main.lean:1]: This change is unrelated and should not be in this PR imo.

**Review by MithicSpirit — CHANGES_REQUESTED** [review]
This currently fails to build (after rebasing on #3).

**Comment by MAO3J1m0Op** [comment]
Now depends on #4.

**Review by MithicSpirit — COMMENTED** [review]
LGTM so far. I will refrain from approving until the dependencies are merged/it stops being a draft and the `Group.inv` notation is agreed upon. It might also be good to not leave the `sorry` in the `mul_inv` definition (IMO either remove it or just prove it properly). I would put a TODO on the `attribute [always_inline]` for whether it's actually doing anything useful or not (once we have more development on the project).

**Review by justin-t-morrill — APPROVED** [review]

**Review by dominicking7 — APPROVED** [review]

File: AlgebraInLean/Basic.lean [hunk 1]
```diff
@@ -1 +1,50 @@
...
+
+  class Monoid (G : Type*) extends Semigroup G where
+    protected id : G
+    protected mul_id : ∀ a : G, a ⋆ id = a
+    protected id_mul : ∀ a : G, id ⋆ a = a
+
+  notation:max "𝕖" => Monoid.id
+
+  @[simp]
+  theorem mul_id [Monoid M] (a : M) : a ⋆ 𝕖 = a := Monoid.mul_id a
```
Comment (AdamJasonKern) [inline AlgebraInLean/Basic.lean]: this should be op_id

File: AlgebraInLean/Basic.lean [hunk 2]
```diff
@@ -1 +1,50 @@
...
+import Mathlib.Tactic
+
+namespace Defs
+  class Magma (G : Type*) where
+    op : G → G → G
+
+  -- FIXME: is this actually helpful/necessary
+  attribute [always_inline] Magma.op
+
+  infixl:75 " ⋆ " => Magma.op
```
Comment (MithicSpirit) [inline AlgebraInLean/Basic.lean]: This character looks somewhat similar to asterisk on some fonts, which is what is usually used for multiplication. It might be worth considering a different character, perhaps as a TODO/FIXME.

**Review by AdamJasonKern — CHANGES_REQUESTED** [review]
We should change the names of the axioms to be more general and not just be limited to multiplication

**Review by MithicSpirit — CHANGES_REQUESTED** [review]

File: AlgebraInLean/Basic.lean (hunk 1 above)
Comment (MithicSpirit) [reply AlgebraInLean/Basic.lean]: I agree, and similarly for the ones below (`id_mul`, `inv_mul`). Remember to also update the definition in `Monoid` above.

**Review by AdamJasonKern — APPROVED** [review]

File: AlgebraInLean/Basic.lean [hunk 3]
```diff
@@ -1 +1,49 @@
-def hello := "world"
\ No newline at end of file
+import Mathlib.Tactic
+
+namespace Defs
+  class Magma (G : Type*) where
+    op : G → G → G
+
+  -- FIXME: is this actually helpful/necessary
+  attribute [always_inline] Magma.op
```
Comment (MithicSpirit) [inline AlgebraInLean/Basic.lean]: I'm pretty sure this doesn't do anything when people use `μ` defined below. I think we should use a `notation` instead of a `def` there, or just be done with it and remove this attribute.

File: AlgebraInLean/Basic.lean [hunk 4]
```diff
@@ -1 +1,49 @@
...
+
+  def μ [Magma G] : G → G → G := Magma.op
+
+  class Semigroup (G : Type*) extends Magma G where
+    protected assoc : ∀ a b c : G, μ (μ a b) c = μ a (μ b c)
+
+  @[simp]
+  theorem assoc [Semigroup G] (a b c : G) : μ (μ a b) c = μ a (μ b c) := Semigroup.assoc a b c
```
Comment (MithicSpirit) [inline AlgebraInLean/Basic.lean]: Maybe call this `op_assoc`? See with others.

**Review by Anoushka1406 — APPROVED** [review]
👍

File: AlgebraInLean/Basic.lean [hunk 5]
```diff
@@ -1 +1,49 @@
...
+
+  @[simp]
+  theorem comm [CommMonoid M] (a b : M) : μ a b = μ b a := CommMonoid.comm a b
+
+  class Group (G : Type*) extends Monoid G where
+    protected inv : G → G
+    protected inv_op : ∀ a : G, μ (inv a) a = 𝕖
+    -- protected op_inv : ∀ a : G, μ a (inv a) = 𝕖
+
+  def ι [Group G] : G → G := Group.inv
```
Comment (MithicSpirit) [inline AlgebraInLean/Basic.lean:39]: If we replace `def μ` with a `notation`, then we should do the same here as well.

File: AlgebraInLean/Basic.lean [hunk 6]
```diff
@@ -1 +1,49 @@
...
+
+  @[simp]
+  theorem assoc [Semigroup G] (a b c : G) : μ (μ a b) c = μ a (μ b c) := Semigroup.assoc a b c
+
+  class Monoid (G : Type*) extends Semigroup G where
+    protected id : G
+    protected op_id : ∀ a : G, μ a id = a
+    protected id_op : ∀ a : G, μ id a = a
+
+  notation:max "𝕖" => Monoid.id
```
Comment (MithicSpirit) [inline AlgebraInLean/Basic.lean]: If we stick with `def μ` then maybe this should be a `def` as well.

**Review by MithicSpirit — CHANGES_REQUESTED** [review]

**Review by 4rim — APPROVED** [review]
LGTM (do note that iota also stands for the inclusion map, though I personally
don't have qualms with it)

**Review by chenne19 — APPROVED** [review]

**Review by william11339 — APPROVED** [review]

**Review by Stavan-Jain — APPROVED** [review]
I don't really like this notation, but it works.
